### PR TITLE
Add separate map for Custom Bridges

### PR DESCRIPTION
**Pull request by karthikeyann**
Summary: Since both Velox bridges, and custom bridges were stored in same map, which did not allow translators to replace/add custom bridges. This change fixes this issue by adding a new map for custom bridges.

A unit test is added to replace the Velox operators with custom operators, and custom bridges, using driver adapter and translator.

Background: To enable running velox operators using libcudf, velox operators will be replaced by custom operators using driver adapter. Bridges are not replaced by driver adapter. But these new operators should also have custom bridges associated with it, to communicate between Build and Probe operators. To create these bridges, we use translator interface provided by velox. Since already a Bridge is stored in a map `bridges` for velox bridges, the translator call does not replace existing bridge and fails silently. This PR separates this map to 2 maps, one for velox join node, another for custom bridges created by translators. This change also helps other driver adaptor implementation (such as wave) to communicate between pipelines using custom bridges.

### CODE REVIEW

File: velox/exec/TaskStructs.h [hunk 1]
```diff
@@ -89,6 +89,8 @@ struct LocalExchangeState {
 struct SplitGroupState {
   /// Map from the plan node id of the join to the corresponding JoinBridge.
   std::unordered_map<core::PlanNodeId, std::shared_ptr<JoinBridge>> bridges;
```
Comment (Yuhta) [inline velox/exec/TaskStructs.h:91]: This can only hold `HashJoinBridge` now, let's update the type to reflect this

File: velox/exec/Task.cpp [hunk 2]
```diff
@@ -1753,7 +1756,8 @@ void Task::addCustomJoinBridgesLocked(
   auto& splitGroupState = splitGroupStates_[splitGroupId];
   for (const auto& planNode : planNodes) {
     if (auto joinBridge = Operator::joinBridgeFromPlanNode(planNode)) {
-      splitGroupState.bridges.emplace(planNode->id(), std::move(joinBridge));
+      splitGroupState.custom_bridges.emplace(
```
Comment (Yuhta) [inline velox/exec/Task.cpp:1759]: Add a check to make sure there is no previous value on the same node ID.  Same change to  `addHashJoinBridgesLocked`.

**Review by Yuhta — COMMENTED** [review]
Can you update a little more background information in summary (e.g. this is for Wave integration of cudf etc.)

File: velox/exec/TaskStructs.h (hunk 1 above)
Comment (karthikeyann) [reply velox/exec/TaskStructs.h:91]: This `bridges` map is used for storing `NestedLoopJoinBridge` too. So, the type needs to be `JoinBridge`.

File: velox/exec/Task.cpp (hunk 2 above)
Comment (karthikeyann) [reply velox/exec/Task.cpp:1759]: Added it to `addHashJoinBridgesLocked`, `addCustomJoinBridgesLocked`, `addNestedLoopJoinBridgesLocked`

File: velox/exec/TaskStructs.h (hunk 1 above)
Comment (Yuhta) [reply velox/exec/TaskStructs.h:91]: Can you add a comment to document that it will contains only `HashJoinBridge` and `NestedLoopJoinBridge`?

Comment (karthikeyann) [reply velox/exec/TaskStructs.h:92]: done.

**Comment by karthikeyann** [comment]
@Yuhta please review again. The CI jobs are not running in this PR.

File: velox/exec/Task.cpp [hunk 3]
```diff
@@ -1742,8 +1745,12 @@ void Task::addHashJoinBridgesLocked(
...
   auto& splitGroupState = splitGroupStates_[splitGroupId];
   for (const auto& planNodeId : planNodeIds) {
-    splitGroupState.bridges.emplace(
-        planNodeId, std::make_shared<HashJoinBridge>());
+    auto const inserted =
+        splitGroupState.bridges
+            .emplace(planNodeId, std::make_shared<HashJoinBridge>())
+            .second;
+    VELOX_CHECK(
+        inserted, "Join bridge for node {} is already present", planNode->id());
```
Comment (karthikeyann) [inline velox/exec/Task.cpp:1753]:
```suggestion
        inserted, "Join bridge for node {} is already present", planNodeId);
```

File: velox/exec/Task.cpp [hunk 4]
```diff
@@ -1762,16 +1775,20 @@ void Task::addCustomJoinBridgesLocked(
...
   auto& splitGroupState = splitGroupStates_[splitGroupId];
   for (const auto& planNodeId : planNodeIds) {
-    splitGroupState.bridges.emplace(
-        planNodeId, std::make_shared<NestedLoopJoinBridge>());
+    auto const inserted =
+        splitGroupState.bridges
+            .emplace(planNodeId, std::make_shared<NestedLoopJoinBridge>())
+            .second;
+    VELOX_CHECK(
+        inserted, "Join bridge for node {} is already present", planNode->id());
```
Comment (karthikeyann) [inline velox/exec/Task.cpp:1791]:
```suggestion
        inserted, "Join bridge for node {} is already present", planNodeId);
```

**Comment by kgpai** [comment]
@karthikeyann Is the test failure related to your change ?

**Comment by karthikeyann** [comment]
Though there are failures in HashJoinTest too, but all the other failures are not due to my change. 
Previous check failures are not due to my change.

**Comment by karthikeyann** [comment]
@kgpai @Yuhta Could you run CI checks again?

**Comment by karthikeyann** [comment]
The logs of 'Sparks Fuzzer' check is almost empty. I can't find the reason for the failure. Any help?
All past 3 failures are from different checks.

**Comment by kgpai** [comment]
@karthikeyann Creating an issue.

**Comment by kgpai** [comment]
Logs are in the summary - downloading them to ascertain why its timing out (potential inf loop ?)

**Comment by kgpai** [comment]
Issue here : https://github.com/facebookincubator/velox/issues/10789 .

Dont see any problems in the logs ; It seems the spark fuzzer just doesnt terminate after running for an hour and just hangs. Anyway the problem is not in this particular PR.

**Pull request closed without merging**